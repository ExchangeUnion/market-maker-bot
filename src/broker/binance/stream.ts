import { Stream } from '../stream';
import WebSocket from 'ws';
import { Logger } from '../../logger';

// If we haven't received a ping during this time we'll consider
// socket dead.
const HEARTBEAT_TIMEOUT = 60 * 1000 * 3 + 1000;
// Since Binance allows to keep a single socket alive for maximum of 24h we're
// recrating the socket after the specified amount of time
const RESTART_TIMEOUT = 60 * 60 * 1000;

class BinanceStream extends Stream {
  private isAlive = false;
  private socket!: WebSocket;
  private pingTimeout!: ReturnType<typeof setTimeout>;
  private restartTimeout!: ReturnType<typeof setTimeout>;

  constructor(private logger: Logger, private tradingPair: string) {
    super();
  }

  public start = (): Promise<void> => {
    return new Promise((resolve) => {
      clearTimeout(this.pingTimeout);
      clearTimeout(this.restartTimeout);
      const url = `wss://stream.binance.com:9443/ws/${this.tradingPair.toLowerCase()}@aggTrade`;
      this.socket = new WebSocket(url);

      this.socket.onopen = () => {
        this.logger.info(`${this.tradingPair} established connection to ${url}`);
      };

      this.socket.on('ping', this.heartbeat);
      this.socket.on('open', this.heartbeat);
      this.socket.on('error', (e) => {
        this.logger.error(`error from the socket ${e}`);
      });

      this.socket.onclose = this.onClose;
      this.socket.onmessage = this.onMessage;
      this.restartTimeout = setTimeout(async () => {
        this.logger.warn(`${this.tradingPair} socket timeout reached - restarting`);
        this.socket.terminate();
        await this.restart();
      }, RESTART_TIMEOUT);
      resolve();
    });
  }

  public close = (): Promise<void> => {
    clearTimeout(this.restartTimeout);
    this.socket.terminate();
    return new Promise((resolve) => {
      const isAliveCheck = setInterval(() => {
        if (!this.isAlive) {
          clearInterval(isAliveCheck);
          resolve();
        }
      });
    });
  }

  private onMessage = (event: WebSocket.MessageEvent) => {
    const aggTrade = JSON.parse(event.data.toString());
    const { p: priceString } = aggTrade;
    const price = parseFloat(priceString);
    this.emit('price', this.tradingPair, price);
  }

  private onClose = (event: WebSocket.CloseEvent) => {
    clearTimeout(this.pingTimeout);
    this.isAlive = false;
    if (event.reason) {
      this.logger.info(`${this.tradingPair} stream closed with reason: ${event.reason}`);
    } else {
      this.logger.info(`${this.tradingPair} stream closed`);
    }
  }

  private restart = async () => {
    try {
      // restart if heartbeat fails
      await this.start();
    } catch (e) {
      this.logger.error(`${this.tradingPair} failed to restart stream: ${e}`);
    }
  }

  private heartbeat = () => {
    this.logger.info(`heartbeat from ${this.tradingPair} socket - isAlive: ${this.isAlive}`);
    this.isAlive = true;
    clearTimeout(this.pingTimeout);
    this.pingTimeout = setTimeout(async () => {
      this.isAlive = false;
      this.socket.terminate();
      await this.restart();
    }, HEARTBEAT_TIMEOUT);
  }

}

export { BinanceStream };
