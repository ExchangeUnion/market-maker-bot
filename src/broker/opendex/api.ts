import {
  ExchangeAPI,
  Balance,
  BalanceProperties,
  CancelOrderRequest,
  QueryOrderResponse,
} from '../api';
import { Logger } from '../../logger';
import { XudGrpcClient } from './xud-client';
import { v4 as uuidv4 } from 'uuid';
import {
  OrderSide as XudOrderSide,
  PlaceOrderResponse,
  SwapSuccess,
 } from './proto/xudrpc_pb';
import { ClientReadableStream, status } from '@grpc/grpc-js';
import { OrderSide } from '../../enums';
import {
  MarketOrderRequest,
  LimitOrderRequest,
  StopLimitOrderRequest,
} from '../exchange';
import { OpenDexOrder } from './order';
import { satsToCoinsStr, coinsToSats } from '../../utils';

class OpenDexAPI extends ExchangeAPI {
  private logger: Logger;
  private certPath: string;
  private rpchost: string;
  private rpcport: number;
  private xudClient!: XudGrpcClient;
  private swapSubscriptions = new Map<string, (swapSuccess: SwapSuccess.AsObject) => unknown >();
  private checkConnectionTimeout: ReturnType<typeof setTimeout> | undefined;

  private swapsCompleteSubscription: ClientReadableStream<SwapSuccess> | undefined;

  constructor(
    { logger, certPath, rpchost, rpcport }:
    {
      logger: Logger,
      certPath: string,
      rpchost: string,
      rpcport: number,
    },
  ) {
    super();
    this.logger = logger;
    this.certPath = certPath;
    this.rpchost = rpchost;
    this.rpcport = rpcport;
  }

  public start = async () => {
    this.xudClient = new XudGrpcClient({
      tlscertpath: this.certPath,
      rpchost: this.rpchost,
      rpcport: this.rpcport,
    });
    this.xudClient.start();
    await this.waitForConnection();
    this.swapsCompleteSubscription = await this.xudClient.subscribeSwaps();
    this.swapsCompleteSubscription.on('data', this.onSwapComplete.bind(this));
    this.swapsCompleteSubscription.on('end', () => {
      this.start();
      // TODO: cancel all orders on other exchanges
    });
    this.swapsCompleteSubscription.on('error', (error: any) => {
      if (error.code && error.code === status.CANCELLED) {
        return;
      } else {
        this.logger.error(`failed to subscribe to xud swaps: ${error}`);
      }
    });
    this.logger.info('OpenDEX API started');
  }

  private waitForConnection = () => {
    return new Promise((resolve, reject) => {
      const verifyConnection = async () => {
        try {
          await this.xudClient.getBalance();
          resolve();
        } catch (e) {
          if (e.code === 14) {
            this.logger.warn('could not verify connection to xud, retrying in 5 sec...');
            this.checkConnectionTimeout = setTimeout(verifyConnection, 5000);
          } else {
            reject();
          }
        }
      };
      verifyConnection();
    });
  }

  public stop = async () => {
    if (this.checkConnectionTimeout) {
      clearTimeout(this.checkConnectionTimeout);
    }
    if (this.swapsCompleteSubscription) {
      this.swapsCompleteSubscription.cancel();
    }
  }

  public getAssets = async (): Promise<Balance[]> => {
    const getBalanceResponse = await this.xudClient.getBalance();
    const assetsMap = new Map<string, BalanceProperties>();
    getBalanceResponse.balancesMap.forEach((balance) => {
      const asset = balance[0];
      const free = parseFloat(
        satsToCoinsStr(balance[1].channelBalance),
      );
      const locked = parseFloat(
        satsToCoinsStr(balance[1].walletBalance),
      );
      assetsMap.set(asset, {
        free,
        locked,
      });
    });
    const tradingLimitsResponse = await this.xudClient.tradingLimits();
    tradingLimitsResponse.limitsMap.forEach((limit) => {
      const asset = limit[0];
      const maxbuy = limit[1].maxbuy;
      const maxsell = limit[1].maxsell;
      let balance = assetsMap.get(asset);
      if (balance) {
        balance = {
          maxbuy,
          maxsell,
          ...balance,
        };
        assetsMap.set(asset, balance);
      }
    });
    const balances: Balance[] = [];
    assetsMap.forEach((balance, asset) => {
      balances.push({
        asset,
        ...balance,
      });
    });
    return balances;
  }

  public newOrder = (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): OpenDexOrder => {
    const orderId = uuidv4();
    const order = new OpenDexOrder({
      ...orderRequest,
      orderId,
      api: this,
      logger: this.logger,
    });
    return order;
  }

  public startOrder = async (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): Promise<PlaceOrderResponse.AsObject> => {
    const pairId = `${orderRequest.baseAsset}/${orderRequest.quoteAsset}`;
    const quantityInSatoshis = coinsToSats(orderRequest.quantity);
    const xudOrderSide = orderRequest.orderSide === OrderSide.Buy
      ? XudOrderSide.BUY
      : XudOrderSide.SELL;
    const xudOrderRequest = {
      pairId,
      orderId: orderRequest.orderId,
      price: orderRequest.price,
      quantity: quantityInSatoshis,
      orderSide: xudOrderSide,
    };
    const placeOrderResponse = await this.xudClient.newOrder(xudOrderRequest);
    return placeOrderResponse;
  }

  public cancelOrder = async (cancelOrderRequest: CancelOrderRequest): Promise<boolean> => {
    try {
      await this.xudClient.removeOrder(cancelOrderRequest.orderId);
      return true;
    } catch (e) {
      return false;
    }
  }

  public subscribeSwap = (id: string, cb: (swapSuccess: SwapSuccess.AsObject) => unknown) => {
    this.swapSubscriptions.set(id, cb);
  }

  public queryOrder = async (): Promise<QueryOrderResponse> => {
    throw new Error('not implemented');
  }

  private onSwapComplete = (swapSuccess: SwapSuccess) => {
    const swapSuccessObject = swapSuccess.toObject();
    const swapSubscription = this.swapSubscriptions.get(swapSuccessObject.localId);
    swapSubscription && swapSubscription(swapSuccessObject);
  }

}

export { OpenDexAPI };
