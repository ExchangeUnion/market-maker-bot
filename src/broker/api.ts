import { OrderSide, OrderType, OrderStatus } from '../enums';
import {
  MarketOrderRequest,
  LimitOrderRequest,
  StopLimitOrderRequest,
} from './exchange';
import { BinanceOrder } from './binance/order';
import { OpenDexOrder } from './opendex/order';

export type BalanceProperties = {
  free: number;
  locked: number;
  maxsell?: number;
  maxbuy?: number;
};

export type Balance = BalanceProperties & {
  asset: string;
};

export type OrderRequest = {
  baseAsset: string;
  quoteAsset: string;
  side: OrderSide;
  type: OrderType;
  quantity: number;
  price?: number;
  stopPrice?: number;
};

export type QueryOrderResponse = {
  tradingPair: string;
  status: OrderStatus;
};

export type QueryOrderRequest = {
  tradingPair: string;
  orderId: string;
};

export type TradingLimits = {
  asset: string;
  maxsell: number;
  maxbuy: number;
};

export type CancelOrderRequest = QueryOrderRequest;

abstract class ExchangeAPI {
  public abstract start(): Promise<void>;
  public abstract stop(): Promise<void>;
  public abstract getAssets(): Promise<Balance[]>;
  public abstract newOrder(
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest
  ): BinanceOrder | OpenDexOrder;
  public abstract queryOrder(
    queryOrderRequest: QueryOrderRequest
  ): Promise<QueryOrderResponse>;
  public abstract cancelOrder(
    cancelOrderRequest: CancelOrderRequest
  ): Promise<boolean>;
}

export { ExchangeAPI };
