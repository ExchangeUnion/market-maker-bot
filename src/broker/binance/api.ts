import assert from 'assert';
import axios from 'axios';
import crypto from 'crypto';
import querystring from 'querystring';
import { OrderSide, OrderType, OrderStatus } from '../../enums';
import { Logger } from '../../logger';
import {
  Balance,
  QueryOrderResponse,
  QueryOrderRequest,
  CancelOrderRequest,
  ExchangeAPI,
} from '../api';
import { v4 as uuidv4 } from 'uuid';
import { LimitOrderRequest, StopLimitOrderRequest } from '../exchange';
import { BinanceOrder } from './order';

const API_BASE_URL = 'https://api.binance.com';
const ACCOUNT_URL = `${API_BASE_URL}/api/v3/account`;
const TEST_ORDER_URL = `${API_BASE_URL}/api/v3/order/test`;
const ORDER_URL = `${API_BASE_URL}/api/v3/order`;
const EXCHANGE_INFO_URL = `${API_BASE_URL}/api/v3/exchangeInfo`;
const AVERAGE_PRICE = `${API_BASE_URL}/api/v3/avgPrice`;
const PING_URL = `${API_BASE_URL}/api/v1/ping`;

type QueryOrderResponseBinance = {
  symbol: string;
  status: OrderStatus;
};

export type OrderFill = {
  commission: string;
  commissionAsset: string;
  price: string;
  qty: string;
  tradeId: number;
};

export type OrderResponse = {
  clientOrderId: string;
  cummulativeQuoteQty: string;
  executedQty: string;
  fills: OrderFill[];
  orderId: number;
  orderListId: number;
  origQty: string;
  price: string;
  side: OrderSide;
  status: string;
  symbol: string;
  timeInForce: string;
  transactTime: number;
  type: string;
};

type BaseFilter = {
  filterType: string;
};

type LotSizeFilter = BaseFilter & {
  minQty: string;
  maxQty: string;
  stepSize: string;
};

type MinNotionalFilter = BaseFilter & {
  minNotional: string;
  applyToMarket: boolean;
  avgPriceMins: number;
};

type Filter = LotSizeFilter | MinNotionalFilter;

type TickerSymbol = {
  baseAsset: string;
  quoteAsset: string;
  filters: Filter[];
};

type ExchangeInfoResponse = {
  symbols: TickerSymbol[];
};

type AvgPriceResponse = {
  sym: string;
  price: string;
};

type RawBalance = {
  asset: string;
  free: string;
  locked: string;
};

type AccountInfo = {
  balances: RawBalance[];
};

require('axios-debug-log')({
  request(debug: any, config: any) {
    debug(`Request with ${config.headers['content-type']}`);
  },
  response(debug: any, response: any) {
    debug('Received HTTP response', response);
    debug(`
      Response with ${response.headers['content-type']},
      from ${response.config.url},
    `);
  },
  error(debug: any, error: any) {
    debug('HTTP error', error);
  },
});

const PING_INTERVAL = 180000;
const EXCHANGE_INFO_INTERVAL = 300000;
const UPDATE_AVERAGE_PRICES_INTERVAL = 300000;

class BinanceAPI extends ExchangeAPI {
  private pingInterval: ReturnType<typeof setInterval> | undefined;
  private exchangeInfoInterval: ReturnType<typeof setInterval> | undefined;
  private updateAveragePricesInterval:
    | ReturnType<typeof setInterval>
    | undefined;
  private exchangeInfo: ExchangeInfoResponse | undefined;
  private averageAssetPrices = new Map<string, number>();
  private tradingPairsToMonitor = new Set<string>();
  private logger: Logger;
  private apiKey: string;
  private apiSecret: string;

  constructor({
    logger,
    apiKey,
    apiSecret,
  }: {
    logger: Logger;
    apiKey: string;
    apiSecret: string;
  }) {
    super();
    this.logger = logger;
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
  }

  public start = async (activeTradingPairs?: Set<string>): Promise<any> => {
    this.pingInterval = setInterval(() => {
      this.ping();
    }, PING_INTERVAL);
    this.exchangeInfoInterval = setInterval(() => {
      this.getExchangeInfo();
    }, EXCHANGE_INFO_INTERVAL);
    if (activeTradingPairs && activeTradingPairs.size > 0) {
      this.tradingPairsToMonitor = activeTradingPairs;
    }
    this.updateAveragePricesInterval = setInterval(() => {
      this.setAveragePrices();
    }, UPDATE_AVERAGE_PRICES_INTERVAL);
    await this.ping();
    await this.getExchangeInfo();
    await this.setAveragePrices();
  };

  public monitorPriceForTradingPair = async (tradingPair: string) => {
    const previousSize = this.tradingPairsToMonitor.size;
    this.tradingPairsToMonitor.add(tradingPair);
    const newSize = this.tradingPairsToMonitor.size;
    if (newSize > previousSize) {
      const avgPriceResponse = await this.getAveragePrice(tradingPair);
      this.averageAssetPrices.set(
        avgPriceResponse.sym,
        parseFloat(avgPriceResponse.price)
      );
    }
  };

  public stopMonitoringPrice = (tradingPair: string) => {
    this.tradingPairsToMonitor.delete(tradingPair);
    this.averageAssetPrices.delete(tradingPair);
  };

  public stop = async () => {
    if (this.pingInterval) {
      clearInterval(this.pingInterval);
    }
    if (this.exchangeInfoInterval) {
      clearInterval(this.exchangeInfoInterval);
    }
    if (this.updateAveragePricesInterval) {
      clearInterval(this.updateAveragePricesInterval);
    }
  };

  public ping = async (): Promise<unknown> => {
    const response = await axios.get(PING_URL);
    assert.equal(response.status, 200);
    return response.data;
  };

  public getExchangeInfo = async (): Promise<
    ExchangeInfoResponse | undefined
  > => {
    try {
      const response = await axios.get(EXCHANGE_INFO_URL);
      assert.equal(response.status, 200);
      this.exchangeInfo = response.data;
      this.logger.info('exchangeInfo updated');
      return response.data;
    } catch (e) {
      this.logger.error(`failed to fetch exchange info ${e}`);
      return;
    }
  };

  public queryOrder = async (
    queryOrderRequest: QueryOrderRequest
  ): Promise<QueryOrderResponse> => {
    const queryOrderResponse = await this.queryOrderInternal(queryOrderRequest);
    return {
      tradingPair: queryOrderResponse.symbol,
      status: queryOrderResponse.status,
    };
  };

  public cancelOrder = async (
    cancelOrderRequest: CancelOrderRequest
  ): Promise<boolean> => {
    try {
      await this.cancelOrderInternal(cancelOrderRequest);
      return true;
    } catch (e) {
      return false;
    }
  };

  private cancelOrderInternal = async (
    cancelOrder: CancelOrderRequest
  ): Promise<void> => {
    const recvWindow = 5000;
    const timestamp = new Date().getTime();
    const { tradingPair, orderId } = cancelOrder;
    const queryString = `timestamp=${timestamp}&recvWindow=${recvWindow}&symbol=${tradingPair}&origClientOrderId=${orderId}`;
    try {
      const response = await axios.delete(
        `${ORDER_URL}?${queryString}&signature=${this.signRequest(
          queryString
        )}`,
        {
          headers: {
            'X-MBX-APIKEY': this.apiKey,
          },
        }
      );
      return response.data;
    } catch (e) {
      if (e.response && e.response.data && e.response.data.msg) {
        this.logger.error(`failed to cancel order: ${e.response.data.msg}`);
      } else {
        this.logger.error(`failed to cancel order: ${e}`);
      }
      return Promise.reject(e);
    }
  };

  private queryOrderInternal = async (
    queryOrder: QueryOrderRequest
  ): Promise<QueryOrderResponseBinance> => {
    if (process.env.LIVE_TRADING) {
      const recvWindow = 5000;
      const timestamp = new Date().getTime();
      const { tradingPair, orderId } = queryOrder;
      const queryString = `timestamp=${timestamp}&recvWindow=${recvWindow}&symbol=${tradingPair}&origClientOrderId=${orderId}`;
      try {
        const response = await axios.get(
          `${ORDER_URL}?${queryString}&signature=${this.signRequest(
            queryString
          )}`,
          {
            headers: {
              'X-MBX-APIKEY': this.apiKey,
            },
          }
        );
        return response.data;
      } catch (e) {
        if (e.response && e.response.data && e.response.data.msg) {
          this.logger.error(
            `failed to query order info: ${e.response.data.msg}`
          );
        } else {
          this.logger.error(`failed to query order info ${e}`);
        }
        return Promise.reject(e);
      }
    } else {
      // Return mock filled response when live trading is disabled
      return {
        symbol: queryOrder.tradingPair,
        status: OrderStatus.Filled,
      };
    }
  };

  public verifyQuantity = async (
    baseAsset: string,
    quoteAsset: string,
    quantity: number
  ) => {
    if (!this.exchangeInfo) {
      throw new Error('exchangeInfo has not been fetched');
    }
    const tradingPair = `${baseAsset}${quoteAsset}`;
    let avgPrice = this.averageAssetPrices.get(tradingPair);
    if (!avgPrice) {
      await this.monitorPriceForTradingPair(tradingPair);
      avgPrice = this.averageAssetPrices.get(tradingPair)!;
    }
    const filters = this.exchangeInfo!.symbols.filter(sym => {
      return sym.baseAsset === baseAsset && sym.quoteAsset === quoteAsset;
    }).map(sym => sym.filters)[0];
    const minNotionalFilter = filters.filter(
      filt => filt.filterType === 'MIN_NOTIONAL'
    )[0] as MinNotionalFilter;
    const minNotional =
      avgPrice * quantity - parseFloat(minNotionalFilter.minNotional);
    if (minNotional <= 0) {
      throw new Error('MIN_NOTIONAL filter failure');
    }
    const lotSize = filters.filter(
      filt => filt.filterType === 'LOT_SIZE'
    )[0] as LotSizeFilter;
    if (
      quantity < parseFloat(lotSize.minQty) ||
      quantity > parseFloat(lotSize.maxQty)
    ) {
      throw new Error('LOT_SIZE filter failure');
    }
    const leftOvers =
      (quantity - parseFloat(lotSize.minQty)) % parseFloat(lotSize.stepSize);
    let resultQty = quantity;
    if (leftOvers) {
      resultQty = resultQty - leftOvers;
    }
    return parseFloat(resultQty.toFixed(8));
  };

  public getAveragePrice = async (sym: string): Promise<AvgPriceResponse> => {
    const response = await axios.get(`${AVERAGE_PRICE}?symbol=${sym}`);
    assert.equal(response.status, 200);
    return { sym, price: response.data.price };
  };

  public accountInfo = async (): Promise<AccountInfo> => {
    const recvWindow = 5000;
    const timestamp = new Date().getTime();
    const queryString = `timestamp=${timestamp}&recvWindow=${recvWindow}`;
    try {
      const response = await axios.get(
        `${ACCOUNT_URL}?${queryString}&signature=${this.signRequest(
          queryString
        )}`,
        {
          headers: {
            'X-MBX-APIKEY': this.apiKey,
          },
        }
      );
      return response.data;
    } catch (e) {
      this.logger.error(`failed to fetch account info ${JSON.stringify(e)}`);
      return { balances: [] };
    }
  };

  public getAssets = async (): Promise<Balance[]> => {
    const accountInfo = await this.accountInfo();
    const parsedBalances = accountInfo.balances.map(balance => {
      return {
        asset: balance.asset,
        free: parseFloat(balance.free),
        locked: parseFloat(balance.locked),
      };
    });
    return parsedBalances;
  };

  public newOrder = (
    orderRequest: LimitOrderRequest | StopLimitOrderRequest
  ): BinanceOrder => {
    const orderId = uuidv4();
    const order = new BinanceOrder({
      ...orderRequest,
      orderId,
      api: this,
      logger: this.logger,
    });
    return order;
  };

  public startOrder = async (
    order: LimitOrderRequest | StopLimitOrderRequest
  ): Promise<string> => {
    const orderResponse = await this.newOrderInternal(order);
    if (process.env.LIVE_TRADING) {
      return orderResponse.clientOrderId;
    } else {
      return uuidv4();
    }
  };

  private newOrderInternal = async (
    order: LimitOrderRequest | StopLimitOrderRequest
  ): Promise<OrderResponse> => {
    try {
      const {
        baseAsset,
        quoteAsset,
        orderSide: side,
        orderType: type,
        quantity,
      } = order;
      const qty = await this.verifyQuantity(baseAsset, quoteAsset, quantity);
      const timestamp = new Date().getTime();
      const tradeData: { [key: string]: string | number } = {
        timestamp,
        side,
        type,
        symbol: `${baseAsset}${quoteAsset}`,
        quantity: qty,
      };
      if (type === OrderType.StopLimit || type === OrderType.Limit) {
        const { price } = order;
        if (!price) {
          return Promise.reject(
            'price is required for limit and stop-limit orders'
          );
        } else {
          tradeData.price = price;
        }
        // Currently all limit and stop-limit
        // orders are good till canceled
        tradeData.timeInForce = 'GTC';
      }
      if (type === OrderType.StopLimit) {
        const { stopPrice } = order as StopLimitOrderRequest;
        if (!stopPrice) {
          return Promise.reject('stopPrice is required for stop-limit orders');
        } else {
          tradeData.stopPrice = stopPrice;
        }
      }
      const queryStringData = querystring.stringify(tradeData);
      let url = `${TEST_ORDER_URL}?signature=${this.signRequest(
        queryStringData
      )}`;
      if (process.env.LIVE_TRADING) {
        url = `${ORDER_URL}?signature=${this.signRequest(queryStringData)}`;
      }
      const response = await axios({
        url,
        method: 'post',
        data: queryStringData,
        headers: {
          'X-MBX-APIKEY': this.apiKey,
          'Content-type': 'application/x-www-form-urlencoded',
        },
      });
      return response.data;
    } catch (e) {
      if (e.response && e.response.data && e.response.data.msg) {
        this.logger.error(`failed to create new order: ${e.response.data.msg}`);
      } else {
        this.logger.error(`failed to create new order ${e}`);
      }
      return Promise.reject(e);
    }
  };

  private setAveragePrices = async () => {
    const averagePricesPromises: Promise<any>[] = [];
    this.tradingPairsToMonitor.forEach(tp => {
      averagePricesPromises.push(this.getAveragePrice(tp));
    });
    const averagePriceResponses = await Promise.all(averagePricesPromises);
    averagePriceResponses.forEach(averagePriceResponse => {
      this.logger.info(
        `updated 5 minute average price for ${averagePriceResponse.sym}: ${averagePriceResponse.price}`
      );
      this.averageAssetPrices.set(
        averagePriceResponse.sym,
        parseFloat(averagePriceResponse.price)
      );
    });
  };

  private signRequest = (query: string): string => {
    return crypto
      .createHmac('sha256', this.apiSecret)
      .update(query)
      .digest('hex');
  };
}

export { BinanceAPI };
