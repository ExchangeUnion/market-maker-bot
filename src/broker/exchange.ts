import {
  ExchangeType,
  OrderType,
  OrderSide,
} from '../enums';
import { Stream } from './stream';
import { BinanceStream } from './binance/stream';
import { OpenDexStream } from './opendex/stream';
import { OpenDexOrder } from './opendex/order';
import { Logger } from '../logger';
import { BinanceAPI } from './binance/api';
import { OpenDexAPI } from './opendex/api';
import { ExchangeAPI, Balance } from './api';
import { BinanceOrder } from './binance/order';

export type OrderRequest = {
  orderId: string,
  baseAsset: string,
  quoteAsset: string,
  orderType: OrderType,
  orderSide: OrderSide,
  quantity: number,
};

export type StopLimitOrderRequest = LimitOrderRequest & {
  stopPrice: number,
};

export type MarketOrderRequest = OrderRequest & {
  price: string,
};

export type LimitOrderRequest = OrderRequest & {
  price: number,
};

class ExchangeBroker {
  private exchange: ExchangeType;
  private priceStreams = new Map<string, Stream>();
  private apiKey: string | undefined;
  private apiSecret: string | undefined;
  private certPath: string | undefined;
  private rpchost: string | undefined;
  private rpcport: number | undefined;
  private logger: Logger;
  private api!: ExchangeAPI;

  constructor(
    {
      exchange,
      apiKey,
      apiSecret,
      certPath,
      logger,
      rpchost,
      rpcport,
    }:
    {
      exchange: ExchangeType,
      logger: Logger,
      apiKey?: string,
      apiSecret?: string,
      certPath?: string,
      rpchost?: string,
      rpcport?: number,
    },
  ) {
    this.exchange = exchange;
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
    this.certPath = certPath;
    this.logger = logger;
    this.rpchost = rpchost;
    this.rpcport = rpcport;
  }

  public start = async () => {
    this.checkConfig();
    if (this.exchange === ExchangeType.Binance) {
      this.api = new BinanceAPI({
        logger: this.logger,
        apiKey: this.apiKey!,
        apiSecret: this.apiSecret!,
      });
    } else if (this.exchange === ExchangeType.OpenDEX) {
      this.api = new OpenDexAPI({
        logger: this.logger,
        certPath: this.certPath!,
        rpchost: this.rpchost!,
        rpcport: this.rpcport!,
      });
    }
    await this.api.start();
  }

  public getAssets = async (): Promise<Balance[]> => {
    const ownedAssets = await this.api.getAssets();
    return ownedAssets;
  }

  public getPrice = async (
    tradingPair: string,
    callback: (tradingPair: string, price: number) => any,
  ): Promise<void> => {
    if (!this.priceStreams.has(tradingPair)) {
      if (this.exchange === ExchangeType.Binance) {
        const binanceStream = new BinanceStream(this.logger, tradingPair);
        this.priceStreams.set(tradingPair, binanceStream);
        binanceStream.on('price', callback);
        await binanceStream.start();
      } else if (this.exchange === ExchangeType.OpenDEX) {
        const openDexStream = new OpenDexStream();
        this.priceStreams.set(tradingPair, openDexStream);
        openDexStream.on('price', callback);
        await openDexStream.start();
      }
    } else {
      const priceStream = this.priceStreams.get(tradingPair)!;
      priceStream.on('price', callback);
      return Promise.resolve();
    }
  }

  public newOrder = (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): BinanceOrder | OpenDexOrder => {
    return this.api.newOrder(orderRequest);
  }

  public close = async () => {
    const priceStreamClosePromises: Promise<any>[] = [];
    this.priceStreams.forEach((priceStream) => {
      priceStreamClosePromises.push(priceStream.close());
    });
    await Promise.all(priceStreamClosePromises);
    if (this.api) {
      await this.api.stop();
    }
  }

  private checkConfig = () => {
    if (
      this.exchange === ExchangeType.Binance &&
      (!this.apiKey || !this.apiSecret)
    ) {
      throw new Error('Binance exchange type requires you to specify the apiKey and apiSecret');
    } else if (
      this.exchange === ExchangeType.OpenDEX &&
      (!this.certPath || !this.rpchost || !this.rpcport)
    ) {
      throw new Error('OpenDEX exchange type requires you to specify the certPath, rpchost and rpcport');
    }
  }
}

export { ExchangeBroker };
