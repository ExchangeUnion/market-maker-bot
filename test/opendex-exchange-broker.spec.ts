import { ExchangeBroker } from '../src/broker/exchange';
import { OpenDexStream } from '../src/broker/opendex/stream';
import { Level, Logger } from '../src/logger';
import { OpenDexAPI } from '../src/broker/opendex/api';
import {
  ExchangeType,
  OrderType,
  OrderSide,
} from '../src/enums';

const loggers = Logger.createLoggers(Level.Warn);

jest.mock('../src/broker/opendex/api');
OpenDexAPI.prototype.start = jest.fn();
OpenDexAPI.prototype.newOrder = jest.fn();

jest.mock('../src/broker/opendex/stream');

describe('ExchangeBroker.OpenDEX', () => {

  describe('newOrder', () => {
    test('it creates a new limit order', async () => {
      const broker = new ExchangeBroker({
        exchange: ExchangeType.OpenDEX,
        certPath: '/path/to/tls.cert',
        rpchost: 'localhost',
        rpcport: 8886,
        logger: loggers.opendex,
      });
      jest.resetAllMocks();
      await broker.start();
      const orderRequest = {
        orderId: '123',
        baseAsset: 'ETH',
        quoteAsset: 'BTC',
        orderType: OrderType.Limit,
        orderSide: OrderSide.Buy,
        quantity: 0.005,
        price: 0.023006,
      };
      await broker.newOrder(orderRequest);
      expect(OpenDexAPI.prototype.newOrder).toHaveBeenCalledTimes(1);
    });
  });

  describe('getPrice', () => {
    let broker: ExchangeBroker;
    const onBTCUSDTpriceChange = jest.fn();
    const startStream = OpenDexStream.prototype.start = jest.fn();
    const onListener = OpenDexStream.prototype.on = jest.fn();

    beforeEach(async () => {
      broker = new ExchangeBroker({
        exchange: ExchangeType.OpenDEX,
        certPath: '/hello/world/tls.cert',
        rpchost: 'localhost',
        rpcport: 8886,
        logger: loggers.opendex,
      });
      await broker.start();
      jest.resetAllMocks();
    });

    test('getPrice creates price stream for tradingPair', async () => {
      expect(broker['priceStreams'].size).toEqual(0);
      const tradingPair = 'BTCUSDT';
      await broker.getPrice(tradingPair, onBTCUSDTpriceChange);
      expect(OpenDexStream).toHaveBeenCalledTimes(1);
      expect(startStream).toHaveBeenCalledTimes(1);
      expect(onListener).toHaveBeenCalledTimes(1);
    });
  });

  describe('verify configuration', () => {

    test('throws without certPath', async () => {
      expect.assertions(1);
      try {
        const broker = new ExchangeBroker({
          exchange: ExchangeType.OpenDEX,
          logger: loggers.opendex,
        });
        await broker.start();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    test('throws without rpchost', async () => {
      expect.assertions(1);
      try {
        const broker = new ExchangeBroker({
          exchange: ExchangeType.OpenDEX,
          logger: loggers.opendex,
          certPath: '/path/to/tls.cert',
        });
        await broker.start();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    test('throws without rpcport', async () => {
      expect.assertions(1);
      try {
        const broker = new ExchangeBroker({
          exchange: ExchangeType.OpenDEX,
          logger: loggers.opendex,
          certPath: '/path/to/tls.cert',
          rpchost: 'localhost',
        });
        await broker.start();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    test('starts with certPath, rpchost and rpcport specified', async () => {
      const broker = new ExchangeBroker({
        exchange: ExchangeType.OpenDEX,
        certPath: '/path/to/tls.cert',
        rpchost: 'localhost',
        rpcport: 8886,
        logger: loggers.opendex,
      });
      jest.resetAllMocks();
      await broker.start();
      expect(OpenDexAPI.prototype.start).toHaveBeenCalledTimes(1);
    });

  });

});
