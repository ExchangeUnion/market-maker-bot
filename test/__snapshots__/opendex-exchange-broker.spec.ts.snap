// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeBroker.OpenDEX verify configuration throws without certPath 1`] = `[Error: OpenDEX exchange type requires you to specify the certPath, rpchost and rpcport]`;

exports[`ExchangeBroker.OpenDEX verify configuration throws without rpchost 1`] = `[Error: OpenDEX exchange type requires you to specify the certPath, rpchost and rpcport]`;

exports[`ExchangeBroker.OpenDEX verify configuration throws without rpcport 1`] = `[Error: OpenDEX exchange type requires you to specify the certPath, rpchost and rpcport]`;
